src/encodings/encoding.rs

```diff
@@ -41,6 +41,23 @@ pub trait Encoder<T: DataType> {
   /// Encodes data from `values`.
   fn put(&mut self, values: &[T::T]) -> Result<()>;
 
+  /// Encodes data from `values`, which contains spaces for null values, that is
+  /// identified by `valid_bits`.
+  ///
+  /// Returns the number of non-null values encoded.
+  fn put_spaced(&mut self, values: &[T::T], valid_bits: &[u8]) -> Result<usize> {
```

Comment:
Yes. However the cost of counting all the bits is non-trivial though, unless we add another parameter to this method. The information on whether values are all not null should be available on the caller side, so whoever calls this can just check that and call `put` instead if that's the case. Let me know your thought.